### PR TITLE
vtgate: Initial implementation of master buffer.

### PR DESCRIPTION
**Pull request by michael-berlin**
The following features will be added in later commits:
- Enforced max failover duration.
- Buffer window i.e. evicting older entries from full buffer.
- Parallelized drain.
- Dry-run mode.

### CODE REVIEW

File: go/vt/vtgate/buffer/buffer.go [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+	// mu guards all fields in this group.
+	// In particular, it is used to serialize the following Go routines:
+	// - 1. Requests which may buffer (RLock, can be run in parallel)
+	// - 2. Request which starts buffering (based on the seen error)
+	// - 3. HealthCheck listener ("StatsUpdate") which stops buffering
+	// - 4. Timer which may stop buffering after -vtgate_buffer_max_failover_duration
+	mu sync.RWMutex
+	// failovers is the list of ongoing failovers.
+	// Key Format: "<keyspace>/<shard>"
+	failovers map[string]*shardBuffer
```
Comment (alainjobart) [inline go/vt/vtgate/buffer/buffer.go]: The three maps you have could be combined into one map, with 3 fields. Or just move externalyReparented and lastFailoverEnd into shardBuffer. That way we only touch the map the first time we see a shard, and keep it after that.

File: go/vt/vtgate/gateway/discoverygateway.go [hunk 2]
```diff
@@ -358,6 +378,33 @@ func (dg *discoveryGateway) withRetry(ctx context.Context, target *querypb.Targe
...
 
 	for i := 0; i < dg.retryCount+1; i++ {
+		// Check if we should buffer MASTER queries which failed due to an ongoing
+		// failover.
+		// Note: We only buffer "!inTransaction" queries i.e.
+		// a) no transaction is necessary (e.g. critical reads) or
+		// b) no transaction was created yet.
+		if !inTransaction && target.TabletType == topodatapb.TabletType_MASTER {
+			// The next call blocks if we should buffer during a failover.
+			if retryDone, bufferErr := dg.buffer.WaitForFailoverEnd(ctx, target.Keyspace, target.Shard, err); bufferErr == nil {
```
Comment (alainjobart) [inline go/vt/vtgate/gateway/discoverygateway.go]: Let's test for bufferErr != nil here:

if retryDone, bufferErr := dg.buffer.WaitForFailoverEnd(ctx, target.Keyspace, target.Shard, err); bufferErr != nil {
  ...
  break
}

... no need to be in brackets here.

File: go/vt/vtgate/buffer/buffer.go (hunk 1 above)
Comment (michael-berlin) [inline go/vt/vtgate/buffer/buffer.go]: Done.

I've changed the code such that the state is kept in `shardBuffer` now and these objects are reused across failovers.

File: go/vt/vtgate/gateway/discoverygateway.go (hunk 2 above)
Comment (michael-berlin) [inline go/vt/vtgate/gateway/discoverygateway.go]: Done.

**Comment by michael-berlin** [comment]
@alainjobart 

Can you please review the latest commits? I would like to wrap up this pull request and then continue with new pull requests along the road.

File: go/vt/vtgate/buffer/shard_buffer.go [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+	bufferCancel func()
+}
+
+func newShardBuffer(keyspace, shard string, bufferSizeSema *sync2.Semaphore) *shardBuffer {
+	return &shardBuffer{
+		keyspace:       keyspace,
+		shard:          shard,
+		bufferSizeSema: bufferSizeSema,
+		statsKey:       []string{keyspace, shard},
+		logTooRecent:   logutil.NewThrottledLogger("FailoverTooRecent", 5*time.Second),
```
Comment (alainjobart) [inline go/vt/vtgate/buffer/shard_buffer.go]: Maybe add keyspace/shard to the name?

**Pull request merged**